src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,266 @@
+
+        :return: UpgradeStatusSpec instance
+        """
+        raise NotImplementedError()
+
+    def upgrade_available(self):
+        """
+        Report on what versions are available to upgrade to
+
+        :return: List of strings
+        """
+        raise NotImplementedError()
+    # <<< from 21 March meeting
+
+    # >>> additional to 21mar: auto-placement of stateful services
+    def add_stateful_service_rule(self, service_type, stateful_service_spec,
+                                  placement_spec):
+        """
+        Stateful service rules serve two purposes:
+         - Optionally delegate device selection to the orchestrator
+         - Enable the orchestrator to auto-assimilate new hardware if it
+           matches the placement spec, without any further calls from ceph-mgr.
+
+        To create a confidence-inspiring UI workflow, use test_stateful_service_rule
+        beforehand to show the user where stateful services will be placed
+        if they proceed.
+        """
+        raise NotImplementedError()
+
+    def test_stateful_service_rule(self, service_type, stateful_service_spec,
+                                   placement_spec):
+        """
+        See add_stateful_service_rule.
+        """
+        raise NotImplementedError()
+
+    def remove_stateful_service_rule(self, service_type, id_):
+        """
+        This will remove the *rule* but not the services that were
+        created as a result.  Those should be converted into statically
+        placed services as if they had been created with add_stateful_service,
+        so that they can be removed with remove_stateless_service
+        if desired.
+        """
+        raise NotImplementedError()
+    # <<<
+
+
+class UpgradeSpec(object):
+    # Request to orchestrator to initiate an upgrade to a particular
+    # version of Ceph
+    def __init__(self):
+        self.target_version = None
+
+
+class UpgradeStatusSpec(object):
+    # Orchestrator's report on what's going on with any ongoing upgrade
+    def __init__(self):
+        self.in_progress = False  # Is an upgrade underway?
+        self.services_complete = []  # Which daemon types are fully updated?
+        self.message = ""  # Freeform description
+
+
+class PlacementSpec(object):
+    """
+    For APIs that need to specify a node subset
+    """
+    def __init__(self):
+        self.label = None
+
+
+class StatefulServiceSpec(object):
+    """
+    Details of stateful service creation.
+
+    For OSDs this should map to what ceph-volume can do.
+    """
+
+    def __init__(self):
+        self.service_type = None  # osd, mon
+
+        self.format = None  # filestore, bluestore (meaningful for OSD only)
+
+        self.hostname = None
+
+        # Map role to device ID:
+        #   Valid keys: data (mon)
+        #   Valid keys: data, journal (filestore)
+        #   Valid keys: data, wal, db (bluestore)
+        self.block_devices = {}
```

Comment:
The idea of not going super-general is to minimise the burden on implementers -- it would be kind of heavy for projects like e.g. Rook to have to implement device-based placement rules for RGW.  If we did want to add that ability, I think it would make more sense as an `extensions` field on StatelessServiceSpec rather than trying to have all services support all types of placement.

The stateful/stateless specs have the key difference that stateful specs are for single daemons, whereas stateless specs describe clusters (max size, min size).  In the case of `add_stateful_service_rule`, the spec object is still describing what individual OSDs look like (probably with a blank `block_devices`), rather than properties of a cluster.

Reply:
I was also throwing this back and forth in my head last night.

> The idea of not going completely general is to minimise the burden on implementers

I think I had a different idea here: Since we are getting an inventory of nodes, we could write some logic in this layer that deals with selecting a subset of them. In this way, neither the Orchestrator python layer nor the orchestrator itself needs to be able to do this filtering.

So if we have an inventory of nodes in this interface, and we have a generalized PlacementSpec, we can generically do the filtering at this layer. A generalized PlacementSpec might look like {hostname: 'ceph-*', label: 'mon', extended: {'has-ssd': true}}

My contrary thought is that this allows the user a lot of freedom to shoot themselves in the foot. Beyond making this easier for us, filtering by label only for stateless (though I think hostname-based could also be useful) is a simple API that maps to a simple UI/CLI above. But, if the user knows that if they want to be that granular, they have to provide labels with that granularity, which would be tedious, and kind of moves the complexity for that case to another place.

Reply:
>  In this way, neither the Orchestrator python layer nor the orchestrator itself needs to be able to do this filtering.

The orchestrator definitely needs to understand the filtering in order to schedule the services as nodes come and go.  When a kubernetes administrator adds a new node and kills another node, it (where 'it' is rook/kubernetes, not ceph-mgr) should be able to migrate an RGW service to the available hardware.  If we use labels, this all "just works", because we told k8s the label, and the label makes it all the way down the stack.

If we required ceph-mgr in the loop to evaluate the mapping of label/regex/etc to physical hosts, then we would need the orchestrator to be able to notify ceph-mgr of new hosts, and receive updates from ceph-mgr about the list of hosts that a placement mapped to.